### PR TITLE
Fix registry mirror issues in packages

### PR DESCRIPTION
**Pull request by tatlat**
*Issue #, if available:*
https://github.com/aws/eks-anywhere/issues/7114

*Description of changes:*
Changes the registry-mirror-cred secret in eksa-packages namespace directly from EKS-Anywhere controller.

*Testing (if applicable):*

*Documentation added/planned (if applicable):*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: config/rbac_default/kustomization.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+- patch: |-
+    - op: replace
+      path: /metadata/name
+      value: eksa-packages-role
+  target:
+    group: rbac.authorization.k8s.io
+    kind: Role
+    name: eksa-manager-role
+    namespace: eksa-packages
+    version: v1
```
Comment (g-gaston) [inline config/rbac_default/kustomization.yaml:38]: Does this mean that every time we add a role for a different namespace we would need to add a couple entries here?

File: controllers/cluster_controller.go [hunk 2]
```diff
@@ -386,6 +388,10 @@ func (r *ClusterReconciler) preClusterProviderReconcile(ctx context.Context, log
 		}
 	}
 
+	if err := r.packagesClient.UpdateSecrets(ctx, r.client, cluster); err != nil {
```
Comment (g-gaston) [inline controllers/cluster_controller.go:391]: could this be part of the `packagesReconcile` method?

File: controllers/cluster_controller_test.go [hunk 3]
```diff
@@ -367,6 +368,7 @@ func TestClusterReconcilerReconcileConditions(t *testing.T) {
 			log := testr.New(t)
 			logCtx := ctrl.LoggerInto(ctx, log)
 
+			mockPkgs.EXPECT().UpdateSecrets(logCtx, testClient, sameName(config.Cluster))
```
Comment (g-gaston) [inline controllers/cluster_controller_test.go:371]: maybe we should think about adding a helper to group all expectations? most tests apparently don't care about the expectations, just about the result

File: pkg/clustermanager/cluster_manager.go [hunk 4]
```diff
@@ -1137,6 +1137,11 @@ func (c *ClusterManager) CreateEKSANamespace(ctx context.Context, cluster *types
 	return c.clusterClient.CreateNamespaceIfNotPresent(ctx, cluster.KubeconfigFile, constants.EksaSystemNamespace)
 }
 
+// CreatePackagesNamespace creates the eksa-packages namespace on the cluster if it doesn't already exist.
+func (c *ClusterManager) CreatePackagesNamespace(ctx context.Context, cluster *types.Cluster) error {
```
Comment (g-gaston) [inline pkg/clustermanager/cluster_manager.go:1141]: curious, why is this needed now?
doesn't the "install packages" functionality (through helm) create this ns?

File: pkg/curatedpackages/packagecontrollerclient.go [hunk 5]
```diff
@@ -236,6 +248,70 @@ func (pc *PackageControllerClient) Enable(ctx context.Context) error {
 	return nil
 }
 
+// UpdateSecrets is used to update the registry-mirror-cred secret used by the packages controller.
+func (pc *PackageControllerClient) UpdateSecrets(ctx context.Context, client client.Client, cluster *anywherev1.Cluster) error {
+	secretName := "registry-mirror-cred"
```
Comment (g-gaston) [inline pkg/curatedpackages/packagecontrollerclient.go:253]: const?

File: pkg/curatedpackages/packagecontrollerclient.go [hunk 6]
```diff
@@ -236,6 +248,70 @@ func (pc *PackageControllerClient) Enable(ctx context.Context) error {
...
+	secretKey := types.NamespacedName{
+		Namespace: constants.EksaPackagesName,
+		Name:      secretName,
+	}
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: constants.EksaPackagesName,
+		},
```
Comment (g-gaston) [inline pkg/curatedpackages/packagecontrollerclient.go:262]: why fill the secret if you are just gonna read with the client?

File: pkg/curatedpackages/packagecontrollerclient.go [hunk 7]
```diff
@@ -236,6 +248,70 @@ func (pc *PackageControllerClient) Enable(ctx context.Context) error {
...
+	}
+	credErr := client.Get(ctx, secretKey, secret)
+	err := fillRegistrySecret(cluster.Name, cluster.Spec.RegistryMirrorConfiguration, secret)
+	if err != nil {
+		return err
+	}
+
+	if apierrors.IsNotFound(credErr) {
```
Comment (g-gaston) [inline pkg/curatedpackages/packagecontrollerclient.go:270]: if you just want to do an "upsert", you don't need to do a get. Just do the upgrade and if the object doesn't exist, it will return a IsNotFound. Then you can call create

However, if you do a `Get` first, at least check if the object needs to be updated or not. That way we avoid an extra write call. Remember that this will be called on every reconcile loop.

File: pkg/curatedpackages/packagecontrollerclient.go [hunk 8]
```diff
@@ -236,6 +248,70 @@ func (pc *PackageControllerClient) Enable(ctx context.Context) error {
...
+		return client.Create(ctx, secret)
+	} else if credErr == nil {
+		return client.Update(ctx, secret)
+	}
+	return credErr
+}
+
+func fillRegistrySecret(clusterName string, registry *anywherev1.RegistryMirrorConfiguration, secret *corev1.Secret) error {
```
Comment (g-gaston) [inline pkg/curatedpackages/packagecontrollerclient.go:278]: how was this secret created before?
I was expecting to have already code that did this. If it's helm, why don't we use helm to update the values as well? this would be duplicated logic in that case
It's possible that I'm not getting this right

File: pkg/curatedpackages/packagecontrollerclient.go [hunk 9]
```diff
@@ -236,6 +248,70 @@ func (pc *PackageControllerClient) Enable(ctx context.Context) error {
...
+
+	dconfig := &dockerConfig{Auths: make(map[string]*dockerAuth)}
+	configData, ok := secret.Data[configName]
+	if ok {
+		err := json.Unmarshal(configData, dconfig)
+		if err != nil {
+			return err
+		}
+	}
+	username, password, err := config.ReadCredentials()
```
Comment (g-gaston) [inline pkg/curatedpackages/packagecontrollerclient.go:296]: why are we reading from env vars here? The controller should avoid at all cost passing data through env vars

File: pkg/workflows/create.go [hunk 10]
```diff
@@ -252,6 +252,12 @@ func (s *CreateWorkloadClusterTask) Run(ctx context.Context, commandContext *tas
 			return &CollectDiagnosticsTask{}
 		}
 
+		err = commandContext.ClusterManager.CreatePackagesNamespace(ctx, workloadCluster)
```
Comment (g-gaston) [inline pkg/workflows/create.go:255]: why wasn't this needed before?

File: config/rbac_default/kustomization.yaml (hunk 1 above)
Comment (tatlat) [reply config/rbac_default/kustomization.yaml:38]: Yes, we would need to patch the name of the role for other namespaces since they never get created if there is already another role with the same name in eksa-system.

File: controllers/cluster_controller.go (hunk 2 above)
Comment (tatlat) [reply controllers/cluster_controller.go:391]: So one of the issues I found was that the controller's `reconcile` method sometimes returns before it reaches `packagesReconcile` which is why I put it here to ensure the secret gets updated.

File: pkg/clustermanager/cluster_manager.go (hunk 4 above)
Comment (tatlat) [reply pkg/clustermanager/cluster_manager.go:1141]: We need the namespace before installing packages or else we can't add role/rolebindings in that ns

File: controllers/cluster_controller.go (hunk 2 above)
Comment (g-gaston) [reply controllers/cluster_controller.go:391]: I see. And what would be the problem on waiting until reconcile doesn't return early?

File: pkg/curatedpackages/packagecontrollerclient.go (hunk 8 above)
Comment (jiayiwang7) [reply pkg/curatedpackages/packagecontrollerclient.go:278]: > it will use the image found in packages bundle instead where it rolls out a new pod. Once that pod rolls out with the new image, the secrets are reset to their default values.

why would the secret be reset to default after pod rolls out with image in package bundle? shouldn't we try solving this issue instead?

**Comment by jiayiwang7** [comment]
mind follow up with a PR to remove `registry-mirror-secret` in package controller if it's no longer used anywhere?
